File: services/29-graphql-yoga/src/auth/context.ts
// GraphQL Yoga context factory — builds the per-request GraphQL context.
//
// Resolution order:
//   1. Authorization: Bearer <JWT> header — verified without a DB hit.
//   2. X-API-Key header — DB lookup via Prisma.
//
// When auth succeeds: context.user is populated with AuthUser.
// When auth fails or is missing: context.user is null.
//   Protected resolvers must check context.user and throw GraphQLError with UNAUTHENTICATED.

import type { YogaInitialContext } from 'graphql-yoga'
import { verifyToken } from '../middleware/auth.js'
import { prisma } from '../db/client.js'
import type { AuthUser } from '../middleware/auth.js'

export interface GraphQLContext {
  user: AuthUser | null
  prisma: typeof prisma
}

export async function buildContext(
  initialContext: YogaInitialContext,
): Promise<GraphQLContext> {
  const req = initialContext.request
  const authHeader = req.headers.get('authorization')
  const apiKey = req.headers.get('x-api-key')

  // --- Attempt 1: Bearer JWT ---
  if (authHeader?.startsWith('Bearer ')) {
    const token = authHeader.slice(7)
    const user = verifyToken(token)
    if (user) {
      return { user, prisma }
    }
  }

  // --- Attempt 2: X-API-Key ---
  if (apiKey) {
    const dbUser = await prisma.user.findUnique({ where: { apiKey } })
    if (dbUser) {
      return {
        user: { id: dbUser.id, email: dbUser.email, name: dbUser.name },
        prisma,
      }
    }
  }

  return { user: null, prisma }
}

File: services/29-strawberry/src/auth/context.py
"""Strawberry GraphQL context factory — builds per-request context.

Resolution order:
  1. Authorization: Bearer <JWT> header — verified without a DB hit.
  2. X-API-Key header — DB lookup via SQLAlchemy.

When auth succeeds: context["user"] is a dict with id, email, name, provider.
When auth fails or is missing: context["user"] is None.
  Protected resolvers must check context["user"] and raise PermissionError.
"""

from typing import Any, Optional

from fastapi import Request
from sqlalchemy.future import select

from src.auth.jwt import verify_token
from src.db.models import User
from src.db.session import AsyncSessionLocal


async def build_context(request: Request) -> dict[str, Any]:
    """Build the Strawberry GraphQL context for a request."""
    auth_header: Optional[str] = request.headers.get("authorization")
    api_key: Optional[str] = request.headers.get("x-api-key")

    # --- Attempt 1: Bearer JWT ---
    if auth_header and auth_header.startswith("Bearer "):
        token = auth_header[len("Bearer "):]
        payload = verify_token(token)
        if payload is not None:
            return {
                "user": {
                    "id": int(payload["sub"]),
                    "email": payload["email"],
                    "name": payload["name"],
                    "provider": payload.get("provider", "github"),
                },
                "request": request,
            }

    # --- Attempt 2: X-API-Key ---
    if api_key:
        async with AsyncSessionLocal() as session:
            result = await session.execute(
                select(User).where(User.api_key == api_key)
            )
            user = result.scalars().first()
        if user is not None:
            return {
                "user": {
                    "id": user.id,
                    "email": user.email,
                    "name": user.name,
                    "provider": user.provider,
                },
                "request": request,
            }

    return {"user": None, "request": request}
